### DISCUSSION TITLE
Questions: what is the dulwich equivalent of git remote -v?

### DISCUSSION THREAD

**Discussion by dclong** [jelmer/dulwich#946] [Q&A] [answered]
I want to get the names of all remotes, which can achieve using `git remote -v`. How can I do the same using dulwich?

**Comment by jelmer** [comment]
There isn't a high level command for this at the moment, but you can inspect the configuration from the repository (which this information can be found) by using ``Repo.get_config_stack()``.

**Comment by dclong** [comment]
Hi Jelmer, 

Thank you very much for your help!

**Comment by jakeogh** [comment, marked as answer]
This is close, I think (push) is just assumed? I don't see any `push` records in the config.

```
$ cat git_remote_v.py 
#!/usr/bin/env python3

import sys
from dulwich.repo import Repo
repo = Repo(sys.argv[1])

for config_file in repo.get_config_stack().backends:
    for k, v in config_file.items():
        #print(f'{k=}', f'{v=}')
        if k[0] == b"remote":
            print(k[1].decode("utf8"), v[b"url"].decode("utf8"))


$ ./git_remote_v.py ~/cfg/_myapps/qcodes
origin git@github.com:jakeogh/Qcodes.git
upstream https://github.com/QCoDeS/Qcodes

$ cd ~/cfg/_myapps/qcodes

$ git remote -v
origin  git@github.com:jakeogh/Qcodes.git (fetch)
origin  git@github.com:jakeogh/Qcodes.git (push)
upstream        https://github.com/QCoDeS/Qcodes (fetch)
upstream        https://github.com/QCoDeS/Qcodes (push)
```

**Comment by jelmer** [comment]
I believe the distinction between fetch and bush comes from the fact that you can configure the remotes differently for those two types of operations. See e.g. https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes
